pkg/controller/controlplane/valuesprovider.go

```diff
@@ -503,13 +495,13 @@ func getConfigChartValues(
 		values["vnetResourceGroup"] = *infraStatus.Networks.VNet.ResourceGroup
 	}
 
-	// Add AvailabilitySet config if the cluster is not zoned.
-	if !infraStatus.Zoned {
-		nodesAvailabilitySet, err := azureapihelper.FindAvailabilitySetByPurpose(infraStatus.AvailabilitySets, apisazure.PurposeNodes)
-		if err != nil {
-			return nil, errors.Wrapf(err, "could not determine availability set for purpose 'nodes'")
-		}
-		values["availabilitySetName"] = nodesAvailabilitySet.Name
+	primaryAvailabilitySet, err := helper.FindAvailabilitySetByPurpose(infraStatus.AvailabilitySets, apisazure.PurposeNodes)
+	if err == nil {
+		values["availabilitySetName"] = primaryAvailabilitySet.Name
+	}
+
+	if helper.IsVMORequired(infraStatus) {
+		values["vmType"] = "vmss"
```

Comment:
Maybe 
```go
	if primaryAvailabilitySet, err := helper.FindAvailabilitySetByPurpose(infraStatus.AvailabilitySets, apisazure.PurposeNodes); err == nil {
		values["availabilitySetName"] = primaryAvailabilitySet.Name
	} else if helper.IsVMORequired(infraStatus) {
		values["vmType"] = "vmss"
	}
```
I think the `else` does add something here, by highlighting that vmos and AVs are mutually exclusive. I would maybe consider placing the `isVMORequired` on top and the AVs in the else clause, because the condition inside the func do take into account the AVs.

Reply:
I moved into a seperate function to only append the respective key if a condition match and return then.